lib/block_view_association.go

```diff
@@ -0,0 +1,1047 @@
+const MaxAssociationValueCharLength int = 256
+const AssociationTypeReservedPrefix = "DESO"
+
+func (bav *UtxoView) _connectCreateUserAssociation(
+	txn *MsgDeSoTxn,
+	txHash *BlockHash,
+	blockHeight uint32,
+	verifySignatures bool,
+) (
+	_totalInput uint64,
+	_totalOutput uint64,
+	_utxoOps []*UtxoOperation,
+	_err error,
+) {
+	// Validate the starting block height.
+	if blockHeight < bav.Params.ForkHeights.AssociationsBlockHeight {
+		return 0, 0, nil, RuleErrorAssociationBeforeBlockHeight
+	}
+
+	// Validate the txn TxnType.
+	if txn.TxnMeta.GetTxnType() != TxnTypeCreateUserAssociation {
+		return 0, 0, nil, fmt.Errorf(
+			"_connectCreateUserAssociation: called with bad TxnType %s", txn.TxnMeta.GetTxnType().String(),
+		)
+	}
+
+	// Connect a basic transfer to get the total input and the
+	// total output without considering the txn metadata.
+	totalInput, totalOutput, utxoOpsForTxn, err := bav._connectBasicTransfer(
+		txn, txHash, blockHeight, verifySignatures,
+	)
+	if err != nil {
+		return 0, 0, nil, errors.Wrapf(err, "_connectCreateUserAssociation: ")
+	}
+	if verifySignatures {
+		// _connectBasicTransfer has already checked that the txn is signed
+		// by the top-level public key, which we take to be the sender's
+		// public key so there is no need to verify anything further.
+	}
+
+	// Grab the txn metadata.
+	txMeta := txn.TxnMeta.(*CreateUserAssociationMetadata)
+
+	// Validate the txn metadata.
+	if err := bav.IsValidCreateUserAssociationMetadata(txn.PublicKey, txMeta); err != nil {
+		return 0, 0, nil, err
+	}
+
+	// At this point, we can assume the metadata is valid: all
+	// PKIDs exist, strings are of an appropriate length, etc.
+
+	// Check if there is an existing matching association entry that will be overwritten.
+	// This existing association entry will be restored if we disconnect this txn.
+	prevAssociationEntry, err := bav.GetUserAssociationByAttributes(txn.PublicKey, txMeta)
+	if err != nil {
+		return 0, 0, nil, errors.Wrapf(err, "_connectCreateUserAssociation: ")
+	}
+	// Delete the existing association entry, if exists.
+	if prevAssociationEntry != nil {
+		bav._deleteUserAssociationEntryMappings(prevAssociationEntry)
+	}
+
+	// Retrieve existing ExtraData to merge with any new ExtraData.
+	prevExtraData := make(map[string][]byte)
+	if prevAssociationEntry != nil {
+		prevExtraData = prevAssociationEntry.ExtraData
+	}
+
+	// Construct new association entry from metadata.
+	currentAssociationEntry := &UserAssociationEntry{
+		AssociationID:    txHash,
+		TransactorPKID:   bav.GetPKIDForPublicKey(txn.PublicKey).PKID,
+		TargetUserPKID:   bav.GetPKIDForPublicKey(txMeta.TargetUserPublicKey.ToBytes()).PKID,
+		AppPKID:          bav._associationAppPublicKeyToPKID(txMeta.AppPublicKey),
+		AssociationType:  txMeta.AssociationType,
+		AssociationValue: txMeta.AssociationValue,
+		ExtraData:        mergeExtraData(prevExtraData, txn.ExtraData),
+		BlockHeight:      blockHeight,
```

Comment:
You're saying don't include the BlockHeight in the association at all? IMO it could be nice for a primitive order by time interface. Can't query or order by it. But you do get it back in the entry responses.

Reply:
Yeah let's leave it. Eventually it may be useful.